FXML-638:  Lower ATen Linear to XTen Linear

cmcgirr-amd opened the pull request:
Introduce the XTen Linear operator as well as the conversion pass for the corresponding.

Part of the work to bring the fully connected layer down to linalg

Commit:
feat(ATenToXten.td): add the xten linear operator and the corresponding conversion function

convert from torch linear to xten linear. Add an additional check to ensure we only receive the correct ranks.

Commit:
test: add two tests to check for expected and non-expected conversion

Commit:
fix(ATenToXTen.td): handle the case where bias is none

the torch linear operator can have the bias set to none, here we handle the
case by simply passing the value along. As our torch->onnx flow will
handle inserting an empty bias tensor. If we get torch mlir from say
torchscript, and it contains a none bias then we pass it along.

In XtenToLinAlg the empty bias is handled and an empty bias tensor is
generated. Following the convention of xten we simply pass along.

Commit:
feat(XTenToLinAlg.cpp): add XTen Linear to LinAlg Linear Conversion

converts directly down to linalg providing the correct conversion types.
also handles when bias is optional or none and creates a empty
bias tensor by default

Commit:
test(xten_to_linalg_linear.mlir): add tests for xten linear to linalg linear

provide tests to check that linear gets converted. Including when bias is none or optional

BraunPhilipp commented on lib/Conversion/ATenToXTenPass.cpp:
```diff
@@ -191,6 +192,16 @@ bool isAdaptiveAvgPoolGlobalAveragePool2D(Value outsizes) {
 }
 
+bool checkLinearForXten(Value input, Value weights, Value bias) {
+  // Ensure that when the bias is given, it is of rank 1
```
Why is the bias always of rank 1 here?

cmcgirr-amd replied:
torch defines their bias in linear to be of rank 1. I just kept the check. They assume you can broadcast the tensor to 2D.

Commit:
feat(ATenVisualGraph.cpp): add support for xten linear in visual graph export

update the supported operators to include linear using the same format as the aten linear operator

Commit:
Merge remote-tracking branch 'origin/feature/fused-ops' into christopher.lower_liner_to_xten_linear.FXML-638

Commit:
Merge remote-tracking branch 'origin/christopher.lower_liner_to_xten_linear.FXML-638' into christopher.convert_xten_to_linalg_linear.FXML-638

Commit:
fix(XTenToLinalg.cpp): use cast instead of dyn_cast

cast uses a isa check to assert we are casting to the correct type

cmcgirr-amd committed:
Merge pull request #24 from ljfitz/christopher.convert_xten_to_linalg_linear.FXML-638

FXML-638: Lower XTen Linear to LinAlg Linear

Commit:
Merge remote-tracking branch 'origin/feature/fused-ops' into christopher.lower_liner_to_xten_linear.FXML-638
